src/App.js

```diff
@@ -1,88 +1,156 @@
+
+  useEffect(() => {
+    setNotCompletedToDos(
+      listOfToDos.filter(todo => todo.completed === false),
+    );
+
+    setToDosToShow(listOfToDos);
+  }, [listOfToDos]);
+
+  function addTodo(event) {
+    if (newToDoTitle.trim().length > 0) {
+      setListOfToDos(prevState => [
+        ...listOfToDos,
+        { title: newToDoTitle, id: +new Date(), completed: false },
+      ]);
+      setNewToDoTitle('');
+    }
+  }
+
+  const changeTodoStatus = (todoId) => {
+    setListOfToDos(listOfToDos.map((todo) => {
+      if (todo.id === todoId) {
+        return { ...todo, completed: !todo.completed };
+      }
+
+      return todo;
+    }));
+  };
+
+  const changeTodoTitle = (todoId, newTitleToSet) => {
+    setListOfToDos(listOfToDos.map((todo) => {
+      if (todo.id === todoId) {
+        return { ...todo, title: newTitleToSet };
+      }
+
+      return todo;
+    }));
+  };
+
+  const changeAllTodosStatus = (statusToSet) => {
+    setListOfToDos(listOfToDos.map(
+      todo => ({ ...todo, completed: statusToSet }),
+    ));
+
+    if (statusToShow !== 'all') {
+      if (statusToSet) {
+        setStatusToShow('completed');
+      } else {
+        setStatusToShow('active');
+      }
+    }
+  };
+
+  useEffect(() => {
+    if (statusToShow === 'all') {
+      setToDosToShow(listOfToDos);
+    } else if (statusToShow === 'active') {
+      setToDosToShow(listOfToDos.filter(
+        todo => todo.completed === false,
+      ));
+    } else {
+      setToDosToShow(listOfToDos.filter(
+        todo => todo.completed === true,
+      ));
+    }
+  }, [statusToShow, listOfToDos]);
+
+  const removeTodo = (todoId) => {
+    setListOfToDos(listOfToDos.filter(todo => todo.id !== todoId));
+  };
 
-function App() {
   return (
-    <section className="todoapp">
-      <header className="header">
-        <h1>todos</h1>
-
-        <form>
-          <input
-            type="text"
-            className="new-todo"
-            placeholder="What needs to be done?"
+    <Context.Provider
+      value={{
+        changeTodoStatus,
+        removeTodo,
+        changeTodoTitle,
+      }}
+    >
+
+      <section className="todoapp">
+        <header className="header">
+          <h1>todos</h1>
+          <form>
+            <input
+              type="text"
+              className="new-todo"
+              placeholder="What needs to be done?"
+              value={newToDoTitle}
+              onChange={event => setNewToDoTitle(event.target.value)}
+              onKeyDown={(event) => {
+                if (event.key === 'Enter') {
+                  event.preventDefault();
+                  addTodo(event);
+                }
+              }}
+            />
+          </form>
+        </header>
+
+        <section className="main">
+
+          {(
+            (notCompletedToDos.length === 0
+              && listOfToDos.length !== 0)
+            || (notCompletedToDos.length === listOfToDos.length
+                && listOfToDos.length !== 0)
```

Comment:
if you really need so long check better to move it somewhere and call it like shouldRenderSomething, but here try to avoid such a  long check